bundles/org.openhab.binding.elroconnects/src/main/java/org/openhab/binding/elroconnects/internal/handler/ElroConnectsBridgeHandler.java

```diff
@@ -0,0 +1,855 @@
+            if (answerContent.length() < 44) {
+                logger.debug("Could not decode answer {}", answerContent);
+                return;
+            }
+            sceneName = (new String(HexUtils.hexToBytes(answerContent.substring(6, 38)))).replaceAll("[@$]*", "");
+            scenes.put(sceneId, sceneName);
+            logger.debug("Scene ID {} name: {}", sceneId, sceneName);
+        }
+    }
+
+    private void processSceneTypeMessage(ElroConnectsMessage message) {
+        String sceneContent = message.getSceneContent();
+        if ("OVER".equals(sceneContent)) {
+            // last message in series received
+            stopAwaitResponse();
+
+            updateSceneOptions();
+        }
+    }
+
+    private void processSceneMessage(ElroConnectsMessage message) {
+        int sceneId = message.getSceneGroup();
+
+        currentScene = sceneId;
+
+        updateState(SCENE, new StringType(String.valueOf(currentScene)));
+    }
+
+    private void processAlarmTriggerMessage(ElroConnectsMessage message) {
+        String answerContent = message.getAnswerContent();
+        if (answerContent.length() < 10) {
+            logger.debug("Could not decode answer {}", answerContent);
+            return;
+        }
+
+        int deviceId = Integer.parseInt(answerContent.substring(6, 10), 16);
+
+        ElroConnectsDeviceHandler handler = deviceHandlers.get(deviceId);
+        if (handler != null) {
+            handler.triggerAlarm();
+        }
+        logger.debug("Device ID {} alarm", deviceId);
+    }
+
+    private @Nullable ElroConnectsDevice addDevice(ElroConnectsMessage message) {
+        int deviceId = message.getDeviceId();
+        String deviceType = message.getDeviceName();
+        ElroDeviceType type = TYPE_MAP.getOrDefault(deviceType, ElroDeviceType.DEFAULT);
+
+        ElroConnectsDevice device;
+        switch (type) {
+            case CO_ALARM:
+            case SM_ALARM:
+            case WT_ALARM:
+            case THERMAL_ALARM:
+                device = new ElroConnectsDeviceGenericAlarm(deviceId, this);
+                break;
+            case CXSM_ALARM:
+                device = new ElroConnectsDeviceCxsmAlarm(deviceId, this);
+                break;
+            case POWERSOCKET:
+                device = new ElroConnectsDevicePowerSocket(deviceId, this);
+                break;
+            case ENTRY_SENSOR:
+                device = new ElroConnectsDeviceEntrySensor(deviceId, this);
+                break;
+            case MOTION_SENSOR:
+                device = new ElroConnectsDeviceMotionSensor(deviceId, this);
+                break;
+            case TH_SENSOR:
+                device = new ElroConnectsDeviceTemperatureSensor(deviceId, this);
+                break;
+            default:
+                logger.debug("Device type {} not supported", deviceType);
+                return null;
+        }
+        device.setDeviceType(deviceType);
+        devices.put(deviceId, device);
```

Comment:
Discovery can use this for devices that have not been created as things.

Reply:
OK, it would be better to keep discovery related code in the discovery class.

Reply:
This is not straightforward. The whole communication loop is inside the bridge, and it is not easy to do the discovery requests independently. While it might be possible to separate it out, I think it adds complexity. So I would prefer keeping it as is.